### PR TITLE
Don't sort in the critical section

### PR DESCRIPTION
**Pull request by wojtek-t**
This is to reduce lock contention in clusters with high number of large list calls.

```release-note
Reduce critical section in watchcache to fix kube-apiserver scalability under heavy load of list requests
```

/kind bug
/priority important-longterm
/sig api-machinery

/assign @jpbetz

### CODE REVIEW

**Comment by sftim** [comment]
If this is a performance improvement, we might want to changelog it.

(I can't tell from the PR description what the main motivation is for merging)

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go [hunk 1]
```diff
@@ -492,21 +492,22 @@ func (s sortableStoreElements) Swap(i, j int) {
...
+func (w *watchCache) WaitUntilFreshAndList(ctx context.Context, resourceVersion uint64, matchValues []storage.MatchValue) (result []interface{}, rv uint64, index string, err error) {
 	if utilfeature.DefaultFeatureGate.Enabled(features.ConsistentListFromCache) && w.notFresh(resourceVersion) {
 		w.waitingUntilFresh.Add()
 		err = w.waitUntilFreshAndBlock(ctx, resourceVersion)
 		w.waitingUntilFresh.Remove()
 	} else {
 		err = w.waitUntilFreshAndBlock(ctx, resourceVersion)
 	}
+
+	defer func() { sort.Sort(sortableStoreElements(result)) }()
```
Comment (mborsz) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go:504]: Readability: Using defer to sort data creates quite non-obvious code flow.

Maybe moving w.waitUnitilFreshAndBlock calls to the anonymous function in line 510 and defer w.RUnlock() there + keeping sort where it was could simplify the flow here?

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go:504]: That was what I did initially, but sorting in defer is more future-proof imho as it ensures that we will not skip it in potential new code paths in the future.

**Comment by wojtek-t** [comment]
> If this is a performance improvement, we might want to changelog it.
> (I can't tell from the PR description what the main motivation is for merging)

We're generally not adding release-notes for perf improvements that are behind the scenes.
I update the PR description to clarify that it's for perf though.

**Review by jpbetz — APPROVED** [review]
/approve
/lgtm

**Pull request merged**